### ISSUE TITLE
Feature Request: A module for Linear CRFs

### ISSUE THREAD

**Issue by RishabhMalviya** [torch/nn#1253]
I've been reading about CRFs (Conditional Random Fields), and it seems they are useful for sequence prediction and pixel-level labeling in images. I'm following [these](https://www.youtube.com/playlist?list=PL6Xpj9I5qXYEcOhn7TqghAJ6NAPrNmUBH) lectures (lectures 3.1 and 3.2 will give you the basic idea).

Basically, linear CRFs are useful if you're trying to label a sequence of input vectors (expecting a sequence of labels as output). The Linear CRF layer will have parameters that have been trained to capture the influence of labels adjacent to each other in the output sequence.

Here's how I think the Linear CRF layer could work:

1. If there are `L` possible labels, the layer would have `L*L` parameters stored in a matrix `M`, with each `M_ij` expressing the likelihood of seeing label `j` after label `i`.
2. If there are `N` vectors in the input sequence, this layer would output `L^N` values, each expressing the likelihood of one of the possible sequences of labels.   

Exactly how the output likelihoods are calculated, is explained in the lectures 3.1 and 3.2 of the linked playlist.

I know a recurrent neural network could probably do sequence labeling anyway, but apparently, CRFs are still popular in some NLP tasks like Named entity Recognition. There are some models I'd like to try out in torch with CRFs, so that's why I'm requesting a CRF layer as a feature. 

Have we ever tried implementing a CRF layer before?
